src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+
+        const declarations: Symbol[] = [];
+
+        // We'll modify these flags as we walk the tree to collect data
+        // about what things need to be done as part of the extraction.
+        let rangeFacts = RangeFacts.None;
+
+        if (!start || !end) {
+            // cannot find either start or end node
+            return { errors: [createFileDiagnostic(sourceFile, span.start, length, Messages.CannotExtractFunction)] };
+        }
+
+        if (start.parent !== end.parent) {
+            // handle cases like 1 + [2 + 3] + 4
+            // user selection is marked with [].
+            // in this case 2 + 3 does not belong to the same tree node
+            // instead the shape of the tree looks like this:
+            //          +
+            //         / \
+            //        +   4
+            //       / \
+            //      +   3
+            //     / \
+            //    1   2
+            // in this case there is no such one node that covers ends of selection and is located inside the selection
+            // to handle this we check if both start and end of the selection belong to some binary operation
+            // and start node is parented by the parent of the end node
+            // if this is the case - expand the selection to the entire parent of end node (in this case it will be [1 + 2 + 3] + 4)
+            const startParent = skipParentheses(start.parent);
+            const endParent = skipParentheses(end.parent);
+            if (isBinaryExpression(startParent) && isBinaryExpression(endParent) && isNodeDescendantOf(startParent, endParent)) {
+                start = end = endParent;
+            }
+            else {
+                // start and end nodes belong to different subtrees
+                return createErrorResult(sourceFile, span.start, length, Messages.CannotExtractFunction);
+            }
+        }
+        if (start !== end) {
+            // start and end should be statements and parent should be either block or a source file
+            if (!isBlockLike(start.parent)) {
+                return createErrorResult(sourceFile, span.start, length, Messages.CannotExtractFunction);
+            }
+            const statements: Statement[] = [];
+            for (const statement of (<BlockLike>start.parent).statements) {
+                if (statement === start || statements.length) {
+                    const errors = checkNode(statement);
+                    if (errors) {
+                        return { errors };
+                    }
+                    statements.push(statement);
+                }
+                if (statement === end) {
+                    break;
+                }
+            }
+            return { targetRange: { range: statements, facts: rangeFacts, declarations } };
+        }
+        else {
+            // We have a single expression (start)

```

Comment:
We check `isStatement` below...

---
In reply to: [132315243](https://github.com/Microsoft/TypeScript/pull/17625#discussion_r132315243) [](ancestors = 132315243)